File: function/nodejs_14_x/src/response/part_number_mapper.ts
import ErrorCode from '../error/error_code';
import RangeResponse from './range_response.types';

const PART_SIZE = 5242880; // 5 MB

export default function mapPartNumber (partNumber: string, transformedObject: Buffer): RangeResponse {
  const objectLength = transformedObject.byteLength;
  const totalParts = Math.ceil(objectLength / PART_SIZE); // part numbers start at 1
  const requestedPart = Number(partNumber);

  if (isNaN(requestedPart) || !Number.isInteger(requestedPart) ||
    requestedPart > totalParts || requestedPart <= 0) {
    return {
      hasError: true,
      errorCode: ErrorCode.INVALID_PART,
      errorMessage: `Cannot specify part number: ${requestedPart}. Use part numbers 1 to ${totalParts}.`
    };
  }

  const partStart = (requestedPart - 1) * PART_SIZE;
  const partEnd = Math.min(partStart + PART_SIZE, objectLength);
  return { object: transformedObject.slice(partStart, partEnd), hasError: false };
}

File: function/java11/src/main/java/com/amazon/s3objectlambda/transform/PartNumberMapper.java
package com.amazon.s3objectlambda.transform;

import com.amazon.s3objectlambda.response.ResponseUtil;
import com.amazon.s3objectlambda.error.Error;

import java.util.Arrays;

/**
 * Handles partNumber requests by applying the partNumber to the transformed object.
 */
public class PartNumberMapper {

    private final Integer partSize = 5242880;

    public TransformedObject mapPartNumber(String partNumber, byte[] responseObjectByteArray) {

        double objectLength;
        TransformedObject responseObject = new TransformedObject();
        objectLength = responseObjectByteArray.length;
        double totalParts = Math.ceil(objectLength / this.partSize);
        int requestedPart;

        try {
            requestedPart = Integer.parseInt(partNumber);
        } catch (NumberFormatException nfe) {
            return ResponseUtil.getErrorResponse("Invalid part number.", Error.INVALID_PART);
        }

        if (requestedPart > totalParts || requestedPart <= 0) {
            return ResponseUtil.getErrorResponse(String.format("Cannot specify part number: %s. " +
                    "Use part number from 1 to %s.", requestedPart, totalParts), Error.INVALID_PART);
        }

        int partStart = (requestedPart - 1) * this.partSize;
        int partEnd = (int) Math.min(partStart + this.partSize, objectLength);
        byte[] objectPart = Arrays.copyOfRange(responseObjectByteArray, partStart, partEnd);

        responseObject.setHasError(false);
        responseObject.setObjectResponse(objectPart);

        return responseObject;
    }
}
